City-of-Helsinki/hel-fi-drupal-grants
# AU-2336: Delete documents from Drupal UI

# [AU-2336](https://helsinkisolutionoffice.atlassian.net/browse/2336)

## What was done

This pull request implements two admin forms that are meant for searching and deleting ATV documents:

- [The Delete testing applications form](https://hel-fi-drupal-grant-applications.docker.so/en/admin/tools/delete-applications): This form can be used to search and delete ATV documents with the following filters: `UUID`, `Business ID`, `appEnv`, `Application type` and `Application status`. The found documents can all be delete by clicking `Delete all`, or a few at a time by clicking `Delete selected`.

- Note that:

  - The form `ONLY` deletes documents that have their status set to `DRAFT`. Any other document are skipped, and their `transaction ID` is printed back to the user.
  
  - Documents with a type of `grants_profile` can only be deleted if `EVERY` other document returned by the search are deleted at the same time.

![Screenshot 2024-04-23 at 10-01-50 Delete testing applications Grants service](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/1540957e-6a4c-432f-86aa-e65fc591cf96)

- [The Delete testing applications by transaction ID form](https://hel-fi-drupal-grant-applications.docker.so/en/admin/tools/delete-by-transaction-id): This form can be used to delete ATV documents by their `transaction ID`, and it accepts a comma separated list of transaction IDs. The form looks up any documents matching the passed in transaction IDs and deletes them, regardless of their status. The intended use-case is to delete all `NON DRAFT` applications from ATV, after they have been deleted from Avus2.

![Screenshot 2024-04-23 at 10 09 38](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/f50b53ce-e7c0-4205-8997-da4ec08ccaff)

- In addition to this, `permission` have been implemented that allow access to each form.

- The forms have been moved under a `tools` menu item.


## How to install

* Make sure your instance is up and running on correct branch.
  * `git checkout feature/AU-2336-delete-documents`
  * `make fresh`
  * Run `make drush-cr`

## How to test

### Test the permissions

- Login with a user that isn't an admin, and make sure they can't access the forms unless they have the permission `grants_admin_delete_applications` or `grants_admin_delete_applications_by_transaction_id`.

### Test the Delete testing applications form

- Login with an admin users and go to the [Delete testing applications form](https://hel-fi-drupal-grant-applications.docker.so/en/admin/tools/delete-applications) form. Play around with the filters and make sure that they work.

- Use the filters `UUID` = `13cb60ae-269a-46da-9a43-da94b980c067` and `appEnv` = `TEST` and click the `Get data` button. Select a few `DRAFT` applications, a few `RECEIVED` applications, and a few `grants_profile` documents. Now click `Delete selected`, and you should be met with an output similar to this:

![Screenshot 2024-04-23 at 10 24 53](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/d185ea04-f870-4221-a0ac-63563f3b1fb5)
```
- Skipped profile deletion: ... (since profiles can only be deleted if EVERY document is deleted).
- The following documents cannot be deleted since they are NOT marked as DRAFT: ... (should output the transaction IDs of the documents that were skipped).
- Processed 3 documents. Deleted documents: 3. ... (a summary of the the documents that were deleted).
- The following documents were deleted: ... (should output the transaction IDs of the deleted documents).
```

- Now use the filters `UUID` = `13cb60ae-269a-46da-9a43-da94b980c067`, `appEnv` = `TEST` and `Application type` = `HYVINYLEIS (68)`. Click the `Delete all above` button, and you should be met with an output similar to this:

![Screenshot 2024-04-23 at 10 37 35](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/9216c2da-7bd0-441d-b657-e4c87c4681d6)

```
- The following documents cannot be deleted since they are NOT marked as DRAFT: ... (should output the transaction IDs of the documents that were skipped).
- Processed 7 documents. Deleted documents: 7. ... (a summary of the the documents that were deleted).
- The following documents were deleted: ... (should output the transaction IDs of the deleted documents).
```

### Test the Delete testing applications by transaction ID form
- Login with an admin users and go to the [Delete testing applications by transaction ID form](https://hel-fi-drupal-grant-applications.docker.so/en/admin/tools/delete-by-transaction-id) form. Paste in a few transaction IDs, like: `TEST-061-0000028, TEST-061-0000029`. Now click `Delete by transaction ID` button, and you should be met with an output similar to this:

![Screenshot 2024-04-23 at 10 44 31](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/73600081-5b9f-458b-aa29-066a8b17b58c)

```
- Processed 2 documents. Deleted documents: 2. ... (a summary of the the documents that were deleted).
- The following documents were deleted: ... (should output the transaction IDs of the deleted documents).
```





[AU-2336]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2336?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
